compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1933,28 +1933,81 @@ TR::Register *OMR::X86::TreeEvaluator::integerDualMulEvaluator(TR::Node *node, T
 
 TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::Register *targetRegister = 0;
-   TR::Node     *firstChild     = node->getFirstChild();
-   TR::Node     *secondChild    = node->getSecondChild();
-   bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   TR::Register *targetRegister  = 0;
+   TR::Node     *firstChild      = node->getFirstChild();
+   TR::Node     *secondChild     = node->getSecondChild();
+   bool          nodeIs64Bit     = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool isByteOperation          = false;
+   bool isMemRef                 = false;
+   TR_X86OpCodes opCode = BADIA32Op, regOpCode = BADIA32Op, memOpCode = BADIA32Op, movOpCode = BADIA32Op;
    intptrj_t    value;
 
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
 
-   if (secondChild->getOpCode().isLoadConst() &&
-      (value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg)))
+   // Opcode evaluation for bmul, smul imul and lmul
+   switch (secondChild->getDataType())
+      {
+      case TR::Int8:
+         value = secondChild->getByte();
+         isMemRef = secondChild->getOpCode().isMemoryReference(); 
+         opCode = isMemRef ? IMUL1AccMem : IMUL1AccReg;
+         regOpCode = IMUL1AccReg;
+         memOpCode = IMUL1AccMem;
+         movOpCode = MOV1RegReg;
+
+         isByteOperation = true;
+         break;
+
+      case TR::Int16:
+         value = secondChild->getShortInt();
+         isMemRef = firstChild->getOpCode().isMemoryReference(); 
+         if (isMemRef)
+            opCode = (value >= -128 && value <= 127) ? IMUL2RegMemImms : IMUL2RegMemImm2;
+         else
+            opCode = (value >= -128 && value <= 127) ? IMUL2RegRegImms : IMUL2RegRegImm2;
+
+         regOpCode = IMUL2RegReg;
+         memOpCode = IMUL2RegMem;
+         movOpCode = MOV2RegReg;
+         break;
+
+      case TR::Int32:
+      case TR::Int64:
+         if(nodeIs64Bit)
+            value = secondChild->getLongInt();
```

Comment:
I still don't see handling to check that the long is representable in 4bytes... Can you explain how you prevent a value outside the range of 4bytes signed is not truncated for encoding into a 4byte immediate in an instruction?